odl/solvers/advanced/douglas_rachford.py

```diff
@@ -0,0 +1,152 @@
+    operators
+
+        ``min_x f(x) + sum_i g_i(L_i x)``
+
+    Where f, g_i are convex functions, L_i are linear `Operator`'s.
+
+    Can also be used to solve the more general problem
+
+        ``min_x f(x) + sum_i (g_i @ l_i)(L_i x)``
+
+    Where l_i are convex functions and  @ is the infimal convolution:
+
+        ``(f @ g)(x) = inf_y { f(x-y) + g(y) }``
+
+    Parameters
+    ----------
+    x : `LinearSpaceVector`
+        Initial point, updated in place.
+    prox_f : `callable`
+        Function returning an `Operator` when called with stepsize.
+        The Operator should be the proximal of ``f``.
+    prox_cc_g : `sequence` of `callable`'s
+        Sequence of functions returning an operator when called with step size.
+        The `Operator` should be the proximal of ``g_i^*``.
+    L : `sequence` of `Operator`'s
+        Sequence of `Opeartor`s` with as many elements as ``prox_cc_gs``.
+    tau : `float`
+        Step size for ``f``.
+    sigma : `sequence` of  `float`
+        Step size for the ``g_i``'s.
+    lam : `float`
+        Step size.
+    niter : `int`
+        Number of iterations.
+    callback : `Callback`, optional
+        Show partial results.
+
+    Other Parameters
+    ----------------
+    prox_cc_l : `sequence` of `callable`'s
+        Sequence of functions returning an operator when called with step size.
+        The `Operator` should be the proximal of ``l_i^*``.
+
+    Notes
+    -----
+    To guarantee convergence, the parameters ``tau``, ``sigma`` and ``L`` need
+    to satisfy
+
+    .. math::
+
+       \\tau \sum_{i=1}^n \\sigma_i ||L_i||^2 < 4
+
+    References
+    ----------
+    For references on the Forward-Backward algorithm, see [BH2013]_.
+    """
+
+    # Problem size
+    m = len(L)
+    if not all(isinstance(op, Operator) for op in L):
+        raise ValueError('`L` not a sequence of operators')
+    if not all(op.is_linear for op in L):
+        raise ValueError('not all operators in `L` are linear')
+    if not all(x in op.domain for op in L):
+        raise ValueError('`x` not in the domain of all operators')
+    if len(sigma) != m:
+        raise ValueError('len(sigma) != len(L)')
+    if len(prox_cc_g) != m:
+        raise ValueError('len(prox_cc_g) != len(L)')
+
+    # Get parameters
+    prox_cc_l = kwargs.pop('prox_cc_l', None)
+    if prox_cc_l is not None and len(prox_cc_l) != m:
+        raise ValueError('`prox_cc_l` not same length as `L`')
+
+    # Check for unused parameters
+    if kwargs:
+        raise TypeError('unexpected keyword argument: {}'.format(kwargs))
+
+    # Pre-allocate values
+    v = [Li.range.zero() for Li in L]
+
+    p1 = x.space.zero()
+    p2 = [Li.range.zero() for Li in L]
+
+    z1 = x.space.zero()
+    z2 = [Li.domain.zero() for Li in L]
+
+    w1 = x.space.zero()
+    w2 = [Li.range.zero() for Li in L]
+
+    for k in range(niter):
+        tmp_1 = sum(Li.adjoint(vi) for Li, vi in zip(L, v))
+        prox_f(tau)(x - (tau / 2.0) * tmp_1, out=p1)
```

Comment:
Could even use `tmp_1.lincomb(1, x, -tau / 2, tmp_1)` and then `prox_f(tau)(tmp_1, out=p1)` here, no?